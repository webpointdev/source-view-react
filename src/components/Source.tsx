import {useMemo, FC, ReactNode} from 'react';
import {LineOfSyntax, SyntaxElement, TreeNode} from 'source-tokenizer';

export type RenderSyntaxTree = (root: TreeNode, defaultRender: RenderSyntaxElement, i: number) => ReactNode;

export type RenderSyntaxElement = (element: SyntaxElement, i: number) => ReactNode;

interface SourceProps {
    source?: string;
    syntax?: LineOfSyntax[];
    renderSyntaxTree?: RenderSyntaxTree;
    widgets?: {[key: number]: ReactNode};
}

const renderGenericElement = (element: SyntaxElement, i: number): ReactNode => {
    if (typeof element === 'string') {
        return element;
    }

    return (
        <span key={i} className={element.properties?.className?.join(' ')}>
            {element.children.map(renderGenericElement)}
        </span>
    );
};

const renderSyntaxContent = (syntax: LineOfSyntax, {renderSyntaxTree}: SourceProps): ReactNode => {
    const render = (element: SyntaxElement, i: number) => {
        if (typeof element === 'string') {
            return element;
        }

        if (renderSyntaxTree) {
            return renderSyntaxTree(element, renderGenericElement, i);
        }

        return renderGenericElement(element, i);
    };

    return syntax.map(render);
};

const reduceLineWith = (props: SourceProps) => {
    const {widgets = {}} = props;

    return (children: ReactNode[], current: string | LineOfSyntax, i: number): ReactNode[] => {
        const lineNumber = i + 1;

        const lineElement = (
            <tr key={`line-${lineNumber}`} className="source-line">
                <td className="source-gutter" data-line-number={lineNumber} />
                <td className="source-code">
                    {typeof current === 'string' ? current : renderSyntaxContent(current, props)}
                </td>
            </tr>
        );
        children.push(lineElement);

        const widget = widgets[lineNumber];
        if (widget) {
            const widgetContainerElement = (
                <tr key={`widget-${lineNumber}`} className="source-widget">
                    <td colSpan={2}>
                        {widget}
                    </td>
                </tr>
            );
            children.push(widgetContainerElement);
        }

        return children;
    };
};

export const Source: FC<SourceProps> = props => {
    const {source, syntax} = props;
    const lines = useMemo(() => source?.split('\n') ?? [], [source]);
    const reduceLine = reduceLineWith(props);

    return (
        <table className="source">
            <tbody>
                {syntax ? syntax.reduce(reduceLine, []) : lines.reduce(reduceLine, [])}
            </tbody>
        </table>
    );
};
