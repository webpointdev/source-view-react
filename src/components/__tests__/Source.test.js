/* eslint-disable react/jsx-no-bind */
import '@testing-library/jest-dom/extend-expect';
import {render, fireEvent} from '@testing-library/react';
import {highlight} from 'refractor';
import {tokenize} from 'source-tokenizer';
import {Source} from '../Source';

const source = `const a = 3;
console.log(3);`;

describe('Source', () => {
    test('text source', () => {
        const {asFragment} = render(<Source source={source} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('syntax source', () => {
        const options = {
            highlight(source) {
                return highlight(source, 'javascript');
            },
        };
        const syntax = tokenize(source, options);
        const {asFragment} = render(<Source source={source} syntax={syntax} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('custom syntax renderer', () => {
        const options = {
            highlight(source) {
                return highlight(source, 'javascript');
            },
        };
        const syntax = tokenize(source, options);
        const stripSyntax = node => {
            if (typeof node === 'string') {
                return node;
            }

            return node.children.map(stripSyntax);
        };
        const {asFragment} = render(<Source syntax={syntax} renderSyntaxTree={stripSyntax} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('with widgets', () => {
        const widgets = {
            2: <p>This is a widget</p>,
        };
        const {getByText} = render(<Source source={source} widgets={widgets} />);
        expect(getByText('This is a widget')).toBeTruthy();
    });

    test('custom class name and style', () => {
        const {asFragment} = render(<Source source="" className="custom" style={{width: 400}} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('custom start line number', () => {
        const {container} = render(<Source source={source} lineStart={8} />);
        expect(container.querySelector('.source-gutter')).toHaveAttribute('data-line-number', '8');
    });

    test('custom gutter event', () => {
        const click = jest.fn();
        const {container} = render(<Source source={source} gutterEvents={{onClick: click}} />);
        fireEvent.click(container.querySelector('.source-gutter'), {});
        expect(click).toHaveBeenCalledTimes(1);
        expect(click.mock.calls[0][0]).toBe(1);
        expect(click.mock.calls[0][1]).toBeTruthy();
    });

    test('custom code event', () => {
        const click = jest.fn();
        const {container} = render(<Source source={source} codeEvents={{onClick: click}} />);
        fireEvent.click(container.querySelector('.source-code'), {});
        expect(click).toHaveBeenCalledTimes(1);
        expect(click.mock.calls[0][0]).toBe(1);
        expect(click.mock.calls[0][1]).toBeTruthy();
    });

    test('custom gutter renderer', () => {
        const renderGutter = line => <span>Custom line {line}</span>;
        const {getByText} = render(<Source source={source} renderGutter={renderGutter} />);
        expect(getByText('Custom line 1')).toBeTruthy();
    });
});
