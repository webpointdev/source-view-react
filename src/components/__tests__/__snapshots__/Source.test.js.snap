// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Source custom class name and style 1`] = `
<DocumentFragment>
  <table
    class="source custom"
    style="width: 400px;"
  >
    <tbody>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter source-gutter-default"
          data-line-number="1"
        />
        <td
          class="source-code"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Source custom syntax renderer 1`] = `
<DocumentFragment>
  <table
    class="source"
  >
    <tbody>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter source-gutter-default"
          data-line-number="1"
        />
        <td
          class="source-code"
        >
          const a = 3;
        </td>
      </tr>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter source-gutter-default"
          data-line-number="2"
        />
        <td
          class="source-code"
        >
          console.log(3);
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Source syntax source 1`] = `
<DocumentFragment>
  <table
    class="source"
  >
    <tbody>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter source-gutter-default"
          data-line-number="1"
        />
        <td
          class="source-code"
        >
          <span
            class="token keyword"
          >
            const
          </span>
           a 
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token number"
          >
            3
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
        </td>
      </tr>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter source-gutter-default"
          data-line-number="2"
        />
        <td
          class="source-code"
        >
          <span
            class="token console class-name"
          >
            console
          </span>
          <span
            class="token punctuation"
          >
            .
          </span>
          <span
            class="token method function property-access"
          >
            log
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token number"
          >
            3
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Source text source 1`] = `
<DocumentFragment>
  <table
    class="source"
  >
    <tbody>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter source-gutter-default"
          data-line-number="1"
        />
        <td
          class="source-code"
        >
          const a = 3;
        </td>
      </tr>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter source-gutter-default"
          data-line-number="2"
        />
        <td
          class="source-code"
        >
          console.log(3);
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
